### ISSUE TITLE
implementation of async functions

### ISSUE THREAD

**Issue by bassamsdata** [ptdewey/pendulum-nvim#4]
Thank you for this plugin. I'm opening this issue following a discussion on [Reddit](https://www.reddit.com/r/neovim/comments/1emqq5d/introducing_pendulum_a_time_tracking_plugin_for/) about implementing async functions instead of sync ones, especially since this is an all-time tracking plugin.

I noticed that you've already switched to using `vim.system()` instead of `vim.fn.system()`. Thank you for that.

Additionally, there are async alternatives for `vim.fn.timer_start`, such as `uv.new_timer()` and `uv.timer_start()`. I'm not sure if there's an async function for `os.time()`, but I saw `uv.gettimeofday() `and `uv.clock_gettime()`. I haven’t used them personally, though.

There's no hurry at all on this. If you're busy, I can take a look at implementing these changes over the weekend if you're okay with PRs.

**Comment by ptdewey** [comment]
I'm likely to be fairly busy for the next few days or so, but I'm totally open to PRs if you are willing to make the changes.
If I do get a chance though, I'll probably look into alternatives to the CursorMoved autocommands.

**Comment by ptdewey** [comment]
Any updates on this? I'm less busy now so if there is anything left I can probably get it done.
Thanks!

**Comment by bassamsdata** [comment]
Thank you for the reminder. I actually refactored the remote file (because the version I had couldn't compile the Go binary), and since then, I've been using my fork and forgot to submit a PR. I saw that you updated the plugin with many async `vim.system` functions. 

But I remember the only thing that made me suggest that is what's happening with every `CursorMoved` event. I'm gonna take a look again at what I've done in the fork version and update here.

Thank you.

**Comment by ptdewey** [comment]
I did experiment a bit locally with using `CursorHold` events instead, but I wasn't able to come up with anything I was happy with. `CursorMoved` isn't ideal, but I the code being executed should be so minimal that I haven't been able to notice any performance impacts (although its possible they are there). 

I'm sure there is a better way of doing it (while also preserving the current logging behavior), so feel to throw out any ideas or suggestions you may have.

**Comment by bassamsdata** [comment]
> CursorMoved isn't ideal, but the code being executed should be so minimal that I haven't been able to notice any performance impacts.

True, I quickly looked into the code, and it's only the `os.time` function, which shouldn't affect the speed, especially in small files. However, in relatively large files, `CursorMovedI` (not `CursorMoved`) might not be ideal because of nvim-cmp and Treesitter are using it as well. But again, I need to profile to see the data.

I'll experiment with the plugin today and let you know the results.

I just want to mention that it's a really cool plugin, and I've been using it ever since.
